File: fix_commit.js
#!/usr/bin/env node
/**
 * Builds a clean commit:
 *  - app/src tree base: 8918cbb (has C4-C7/C8/Nava badge changes)
 *  - app/src/main.jsx replaced with clean blob from 2f9c818
 *  - app/src/App.jsx  replaced with complete blob 3bf875f8
 *  - root vercel.json updated for Root Dir = app
 */
const { spawnSync } = require('child_process');
const fs   = require('fs');
const path = require('path');

const REPO = __dirname;

function git(...args) {
  const r = spawnSync('git', args, { cwd: REPO, encoding: 'utf8' });
  if (r.status !== 0) { console.error(`ERROR: git ${args.join(' ')}\n${r.stderr}`); process.exit(1); }
  return r.stdout.trim();
}

const TMP = path.join(REPO, '.git', '_mktree_input.txt');
function mktree(rows) {
  const input = rows.map(r => `${r.mode} ${r.type} ${r.sha}\t${r.name}`).join('\n') + '\n';
  fs.writeFileSync(TMP, input, 'utf8');
  const fd = fs.openSync(TMP, 'r');
  const r = spawnSync('git', ['mktree'], { cwd: REPO, encoding: 'utf8', stdio: [fd, 'pipe', 'pipe'] });
  fs.closeSync(fd);
  if (r.status !== 0) { console.error(`mktree failed:\n${r.stderr}\nInput:\n${input}`); process.exit(1); }
  return r.stdout.trim();
}

function lsTree(sha) {
  return git('ls-tree', sha).split('\n').filter(Boolean).map(line => {
    const tab  = line.indexOf('\t');
    const name = line.slice(tab + 1);
    const [mode, type, blobSha] = line.slice(0, tab).split(' ');
    return { mode, type, sha: blobSha, name };
  });
}

// ── Known-clean blobs (src/ level) ────────────────────────────────────────
const CLEAN_SRC = {
  'main.jsx': '6603ddb0bbb07476fdb2d02044d9436f37d7df9e', // complete, from 2f9c818
  'App.jsx':  '3bf875f808dfefb4ac4e3f7d0c7785b3d1e599a0', // complete, ends with </aside>
};

// ── Known-clean blobs (src/components/ level) ─────────────────────────────
const CLEAN_COMP = {
  // C4-C7: SVG inline arrows restored (text badges removed)
  'C4View.jsx': 'd974cceffc172e36a84816a18b2b0d10545d8170',
  'C5View.jsx': '7d46560311bfcdb3fdd898a97ff8843a9483a88e',
  'C6View.jsx': '2e71969767ad1f84cd4afc2f625b6582f1a0ac79',
  'C7View.jsx': '420894fd9b6a7550896c2eb15d41476fe655d05e',
  // C8View: complete pre-badge blob + ↻ clockwise badge patched in sandbox
  'C8View.jsx': 'b778b122fcddbbd72e6f19edf847f215f3fb6e16',
  // NavaChakreshvariView: complete + mobile explore instruction patched in sandbox
  'NavaChakreshvariView.jsx': '57337b122c1f32e4c6c4d9b1903fda0dae929b75',
};

// Root vercel.json for Root Dir = ./ : cd app && npm install && npm run build
const ROOT_VERCEL = '7849c518f3239591506a7167ccacf877d3fa77ed';

// src tree from 8918cbb (has all intended C4-C7/C8/Nava badge changes)
const BASE_SRC_TREE = 'f0b94b0d4ae4d3869dfc9e64efd5e096bcb07dd8';

// ── Verify ─────────────────────────────────────────────────────────────────
for (const [name, sha] of [...Object.entries(CLEAN_SRC), ...Object.entries(CLEAN_COMP)]) {
  const t = git('cat-file', '-t', sha);
  if (t !== 'blob') { console.error(`MISSING ${sha} (${name})`); process.exit(1); }
  console.log(`  ok  ${sha}  ${name}`);
}
if (git('cat-file', '-t', BASE_SRC_TREE) !== 'tree') { console.error('Missing base src tree'); process.exit(1); }
console.log(`  ok  ${BASE_SRC_TREE}  base src tree (8918cbb)`);
if (git('cat-file', '-t', ROOT_VERCEL) !== 'blob') { console.error('Missing root vercel.json blob'); process.exit(1); }
console.log(`  ok  ${ROOT_VERCEL}  root vercel.json`);

// ── Rebuild components tree ─────────────────────────────────────────────────
const compSha    = lsTree(BASE_SRC_TREE).find(r => r.name === 'components').sha;
const newCompTree = mktree(lsTree(compSha).map(r =>
  CLEAN_COMP[r.name] ? { ...r, sha: CLEAN_COMP[r.name] } : r
));
console.log(`new comp  = ${newCompTree}`);

// ── Rebuild src tree ────────────────────────────────────────────────────────
const newSrcTree = mktree(lsTree(BASE_SRC_TREE).map(r => {
  if (r.name === 'components') return { ...r, sha: newCompTree };
  if (CLEAN_SRC[r.name])       return { ...r, sha: CLEAN_SRC[r.name] };
  return r;
}));
console.log(`\nnew src   = ${newSrcTree}`);

// ── HEAD info ───────────────────────────────────────────────────────────────
const headSha   = git('rev-parse', 'HEAD');
const rootTree  = git('cat-file', '-p', 'HEAD').split('\n')[0].replace('tree ', '').trim();
console.log(`HEAD      = ${headSha}`);
console.log(`root tree = ${rootTree}`);

// ── Rebuild app tree ────────────────────────────────────────────────────────
const appSha    = lsTree(rootTree).find(r => r.name === 'app').sha;
// app/vercel.json must be present for Root Dir = app builds
const APP_VERCEL = '0654b16c1a6f399628e64041d9ce6d62737fefb7';
// index.html with build comment to force a real tree diff
const INDEX_HTML = 'cb77e0f1db4b21b8e99d34a91de3aafb7050345d'; // force-v5
const appRows = lsTree(appSha).map(r => {
  if (r.name === 'src')        return { ...r, sha: newSrcTree };
  if (r.name === 'index.html') return { ...r, sha: INDEX_HTML };
  return r;
});
// Add app/vercel.json if missing
if (!appRows.find(r => r.name === 'vercel.json')) {
  appRows.push({ mode: '100644', type: 'blob', sha: APP_VERCEL, name: 'vercel.json' });
  appRows.sort((a, b) => a.name.localeCompare(b.name));
}
const newAppTree = mktree(appRows);
console.log(`new app   = ${newAppTree}`);

// ── Rebuild root tree ───────────────────────────────────────────────────────
const newRootTree = mktree(lsTree(rootTree).map(r => {
  if (r.name === 'app') return { ...r, sha: newAppTree };
  if (r.name === 'vercel.json') return { ...r, sha: ROOT_VERCEL };
  return r;
}));
console.log(`new root  = ${newRootTree}`);

// ── Commit ──────────────────────────────────────────────────────────────────
const MSG = 'fix: clean src from 8918cbb + fixed main.jsx + complete App.jsx';
const cr  = spawnSync('git', ['commit-tree', newRootTree, '-p', headSha, '-F', '-'], {
  cwd: REPO, encoding: 'utf8', input: MSG + '\n'
});
if (cr.status !== 0) { console.error(cr.stderr); process.exit(1); }
const newCommit = cr.stdout.trim();
console.log(`new commit = ${newCommit}`);

// ── Update refs ─────────────────────────────────────────────────────────────
const branch  = git('rev-parse', '--abbrev-ref', 'HEAD');
const refsDir = path.join(REPO, '.git', 'refs', 'heads');
fs.writeFileSync(path.join(refsDir, branch), newCommit + '\n');
console.log(`refs/heads/${branch} -> ${newCommit}`);
if (branch !== 'master') {
  const mRef = path.join(refsDir, 'master');
  if (fs.existsSync(mRef)) { fs.writeFileSync(mRef, newCommit + '\n'); console.log(`refs/heads/master -> ${newCommit}`); }
}

console.log('\nDone. Run:\n  git push origin HEAD:master --force');

File: fix_commit.py
#!/usr/bin/env python3
"""
Builds a clean commit from known-good blob SHAs, bypassing disk reads.
Run from the repo root: python fix_commit.py
"""
import subprocess, sys, os

REPO = os.path.dirname(os.path.abspath(__file__))

def git(*args, input=None):
    result = subprocess.run(
        ["git"] + list(args),
        cwd=REPO,
        capture_output=True,
        text=True,
        input=input,
    )
    if result.returncode != 0:
        print(f"ERROR: git {' '.join(args)}\n{result.stderr}")
        sys.exit(1)
    return result.stdout.strip()

# ── 1. Verify clean blobs exist ──────────────────────────────────────────────
CLEAN_BLOBS = {
    "C4View.jsx":             "ac1ca928",
    "C5View.jsx":             "400fe519",
    "C6View.jsx":             "117cfa08",
    "C7View.jsx":             "2e334005",
    "C8View.jsx":             "c2295b51",
    "NavaChakreshvariView.jsx": "077abde2",
}

print("Verifying clean blobs...")
for name, sha in CLEAN_BLOBS.items():
    t = git("cat-file", "-t", sha)
    if t != "blob":
        print(f"MISSING blob {sha} for {name}. Aborting.")
        sys.exit(1)
    print(f"  ✓  {sha}  {name}")

# ── 2. Read current HEAD tree ────────────────────────────────────────────────
head_sha = git("rev-parse", "HEAD")
print(f"\nHEAD = {head_sha}")

def ls_tree(tree_sha):
    """Return list of (mode, type, sha, name) for a tree."""
    out = git("ls-tree", tree_sha)
    rows = []
    for line in out.splitlines():
        meta, name = line.split("\t", 1)
        mode, typ, sha = meta.split()
        rows.append((mode, typ, sha, name))
    return rows

# ── 3. Rebuild app/src/components tree ──────────────────────────────────────
def get_subtree(path_parts, tree_sha):
    """Walk down path_parts, returning the tree SHA at the leaf."""
    for part in path_parts:
        rows = ls_tree(tree_sha)
        for mode, typ, sha, name in rows:
            if name == part and typ == "tree":
                tree_sha = sha
                break
        else:
            print(f"Could not find subtree '{part}' in {tree_sha}")
            sys.exit(1)
    return tree_sha

root_tree = git("rev-parse", "HEAD^{tree}")
app_tree     = get_subtree(["app"], root_tree)
src_tree     = get_subtree(["src"], app_tree)
comp_tree    = get_subtree(["components"], src_tree)

print(f"\ncomponents tree = {comp_tree}")

# Build updated components tree
comp_rows = ls_tree(comp_tree)
new_comp_lines = []
for mode, typ, sha, name in comp_rows:
    if name in CLEAN_BLOBS:
        new_sha = CLEAN_BLOBS[name]
        print(f"  replacing {name}: {sha} → {new_sha}")
        new_comp_lines.append(f"{mode} {typ} {new_sha}\t{name}")
    else:
        new_comp_lines.append(f"{mode} {typ} {sha}\t{name}")

new_comp_tree = git("mktree", input="\n".join(new_comp_lines) + "\n")
print(f"new components tree = {new_comp_tree}")

# ── 4. Rebuild src tree ──────────────────────────────────────────────────────
src_rows = ls_tree(src_tree)
new_src_lines = []
for mode, typ, sha, name in src_rows:
    if name == "components":
        new_src_lines.append(f"{mode} {typ} {new_comp_tree}\t{name}")
    else:
        new_src_lines.append(f"{mode} {typ} {sha}\t{name}")

new_src_tree = git("mktree", input="\n".join(new_src_lines) + "\n")
print(f"new src tree = {new_src_tree}")

# ── 5. Rebuild app tree ──────────────────────────────────────────────────────
app_rows = ls_tree(app_tree)
new_app_lines = []
for mode, typ, sha, name in app_rows:
    if name == "src":
        new_app_lines.append(f"{mode} {typ} {new_src_tree}\t{name}")
    else:
        new_app_lines.append(f"{mode} {typ} {sha}\t{name}")

new_app_tree = git("mktree", input="\n".join(new_app_lines) + "\n")
print(f"new app tree = {new_app_tree}")

# ── 6. Rebuild root tree ─────────────────────────────────────────────────────
root_rows = ls_tree(root_tree)
new_root_lines = []
for mode, typ, sha, name in root_rows:
    if name == "app":
        new_root_lines.append(f"{mode} {typ} {new_app_tree}\t{name}")
    else:
        new_root_lines.append(f"{mode} {typ} {sha}\t{name}")

new_root_tree = git("mktree", input="\n".join(new_root_lines) + "\n")
print(f"new root tree = {new_root_tree}")

# ── 7. Create commit ─────────────────────────────────────────────────────────
new_commit = git(
    "commit-tree", new_root_tree,
    "-p", head_sha,
    "-m", "fix: clean C4-C7 C8 Nava blobs, replace SVG arrows with text badges",
)
print(f"\nnew commit = {new_commit}")

# ── 8. Update local refs ─────────────────────────────────────────────────────
# Update both main and master to point to new commit
current_branch = git("rev-parse", "--abbrev-ref", "HEAD")
print(f"current branch = {current_branch}")

# Write the ref for current branch
with open(os.path.join(REPO, ".git", "refs", "heads", current_branch), "w") as f:
    f.write(new_commit + "\n")
print(f"Updated refs/heads/{current_branch} → {new_commit}")

# Also update master if it's different
if current_branch != "master":
    master_ref = os.path.join(REPO, ".git", "refs", "heads", "master")
    if os.path.exists(master_ref):
        with open(master_ref, "w") as f:
            f.write(new_commit + "\n")
        print(f"Updated refs/heads/master → {new_commit}")

print("\nDone. Now run:")
print("  git push origin HEAD:master --force")
